napari/utils/events/evented_model.py

```diff
@@ -55,15 +71,112 @@ def no_class_attributes():
     def _return2(x, y):
         return y
 
-    main.ClassAttribute = _return2
+    pydantic_main.ClassAttribute = _return2
     try:
         yield
     finally:
         # undo our monkey patch
-        main.ClassAttribute = utils.ClassAttribute
+        pydantic_main.ClassAttribute = pydantic_utils.ClassAttribute
 
 
-class EventedMetaclass(main.ModelMetaclass):
+class ParametrizedGenericCompliantModelField(ModelField):
+    """
+    A ModelField that correcly interprets and validates a wider variety
+    of parametrized generics.
+
+    See:
+    - https://github.com/napari/napari/pull/4609
+    - https://github.com/samuelcolvin/pydantic/issues/4161
+    """
+
+    def _type_analysis(self):
+        super()._type_analysis()
+        # if class validators are unset, it means pydantic ignored them, so
+        # we add them back in just like pydantic does
+        if not self.class_validators:
+            get_validators = getattr(
+                self.outer_type_, '__get_validators__', dict
+            )
+            self.class_validators.update(
+                {
+                    f'{self.outer_type_.__name__}_{i}': Validator(
+                        validator, pre=True
+                    )
+                    for i, validator in enumerate(get_validators())
+                }
+            )
+
+        origin = get_origin(self.outer_type_)
+        # since issubclass(collections.abc.Set, typing.Set) == False, and our EventedSet
+        # is a collections.abc.Set, we have to tell pydantic to treat it the same as the other sets
+        # isinstance(origin, type) is needed for non-class origins (e.g: Literal)
+        if (
+            origin is not None
+            and isinstance(origin, type)
+            and issubclass(origin, abc.Set)
+        ):
+            self.shape = SHAPE_SET
+            self.type_ = get_args(self.outer_type_)[0]
+
+    def populate_validators(self):
+        super().populate_validators()
+        # we need to get rid of the basic arbitrary_types validator because otherwise
+        # in cases where we don't coerce type (e.g: Colormap) it always fails cause the type is wrong
+        # We then coerce the type ourselves below
+        self.validators = [
+            f
+            for f in self.validators
+            if f.__name__ != 'arbitrary_type_validator'
+        ]
+
+        origin = get_origin(self.outer_type_) or self.outer_type_
+        # anything that's not really a class should not be used for coercion
+        if (
+            not isinstance(origin, type)
+            or origin is Union
+            or issubclass(origin, Protocol)
```

Comment:
if `origin is Union` then isn't it by definition not an instance of `type`? 
```python
In [20]: isinstance(Union, type)
Out[20]: False
```

also, why can't a Protocol subclass be used for coercion?

```python
In [21]: class P(Protocol):
    ...:     def x(self): ...
    ...:

In [22]: class C(P):
    ...:     def x(self): return 1
    ...:

In [23]: C()  # we could certainly call this origin and return an instance
Out[23]: <__main__.C object at 0x13db53f10>

In [24]: issubclass(C, Protocol)
Out[24]: True
```

The difference between `P` and `C` here is:
```python
In [27]: C._is_protocol
Out[27]: False

In [28]: P._is_protocol
Out[28]: True
```

Reply:
The problem here is `issubclass(A, B)` raising an exception if `A` is not an instance of `type`. I mixed a few checks together so this became a bit confusing (and maybe wrong). The `Protocol` check was added after I had some issue, but maybe those are solved now... wil check!

Reply:
Renmoving protocol will fail with `Colormap`... If you can check that out it would be great! Just remove that line and run `pytest` :P

Reply:
> The problem here is issubclass(A, B) raising an exception if A is not an instance of type

right, you definitely need the `isinstance(origin, type)` check... but I'm saying the `is Union` check should be removed (it will always be False if origin is an instance of `type`), and the protocol check could be `getattr(origin, '_is_protocol', False)`